### PR TITLE
Warn on unrecognized fields in UIFormat

### PR DESCRIPTION
**Pull request by msteiger**
Logs a warning if unknown json entries are encountered during deserialization of `.ui` files.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/205/
Hooray Jenkins reported success with all tests good!

File: engine/src/main/resources/assets/ui/menu/selectModsScreen.ui [hunk 1]
```diff
@@ -123,7 +123,7 @@
                                     "type" : "ColumnLayout",
                                     "id" : "simpleItems",
                                     "columns" : 2,
-                                    "columnWidths" : [0.3, 0.7],
+                                    "column-widths" : [0.3, 0.7],
```
Comment (Cervator) [inline engine/src/main/resources/assets/ui/menu/selectModsScreen.ui]: Curious: why the change here?

**Comment by Cervator** [comment]
Tested this out, caught a couple examples with JoshariasSurvival (ping @Josharias !):

```
22:12:40.742 [main] WARN  o.t.rendering.nui.asset.UIFormat - Field 'emptyIcon' not recognized for interface org.terasology.rendering.nui.UIWidget in {"type":"UIIconBar","id":"breathBar","icon":"engine:icons#bubble","family":"breathBar","emptyIcon":"engine:icons#burstBubble","halfIconMode":"shrink","spacing":2,"maxIcons":10,"layoutInfo":{"use-content-width":true,"use-content-height":true,"position-left":{"target":"CENTER"},"position-right":{"target":"RIGHT","widget":"toolbar"},"position-bottom":{"target":"TOP","widget":"toolbar","offset":1}}}
22:12:40.743 [main] WARN  o.t.rendering.nui.asset.UIFormat - Field 'crosshairIcon' not recognized for interface org.terasology.rendering.nui.UIWidget in {"type":"UICrosshair","id":"crosshair","crosshairIcon":"engine:gui#crosshair","chargeStages":["engine:gui#crosshairCharge1","engine:gui#crosshairCharge2","engine:gui#crosshairCharge3","engine:gui#crosshairCharge4","engine:gui#crosshairCharge5","engine:gui#crosshairCharge6","engine:gui#crosshairCharge7","engine:gui#crosshairCharge8"],"layoutInfo":{"use-content-width":true,"use-content-height":true,"position-horizontal-center":{},"position-vertical-center":{}}}
```

Also checked out Master of Oreon since it has some custom (and aging) UI stuff, plus it relies on Miniion which is even older (and more broken). Was surprised to see its UI actually work *better* than I remember - I don't recall seeing actual resource counts on the design pane! The Oreon summoning works too, although the King disappears almost immediately again for some reason.

Anyway it did show another couple UI warnings:

```
22:18:51.815 [main] WARN  o.t.rendering.nui.asset.UIFormat - Field 'emptyIcon' not recognized for interface org.terasology.rendering.nui.UIWidget in {"type":"UIIconBar","id":"breathBar","icon":"engine:icons#bubble","family":"breathBar","emptyIcon":"engine:icons#burstBubble","halfIconMode":"shrink","spacing":2,"maxIcons":10,"layoutInfo":{"use-content-width":true,"use-content-height":true,"position-left":{"target":"CENTER"},"position-right":{"target":"RIGHT","widget":"toolbar"},"position-bottom":{"target":"TOP","widget":"toolbar","offset":1}}}
22:18:51.817 [main] WARN  o.t.rendering.nui.asset.UIFormat - Field 'crosshairIcon' not recognized for interface org.terasology.rendering.nui.UIWidget in {"type":"UICrosshair","id":"crosshair","crosshairIcon":"engine:gui#crosshair","chargeStages":["engine:gui#crosshairCharge1","engine:gui#crosshairCharge2","engine:gui#crosshairCharge3","engine:gui#crosshairCharge4","engine:gui#crosshairCharge5","engine:gui#crosshairCharge6","engine:gui#crosshairCharge7","engine:gui#crosshairCharge8"],"layoutInfo":{"use-content-width":true,"use-content-height":true,"position-horizontal-center":{},"position-vertical-center":{}}}
```

I wonder how @mkienenb and @overdhose are doing, sure been a while, alas :-)

**Comment by Josharias** [comment]
Such a great idea.  I have probably copy-pasted failure all over the place.  Thanks msteiger!

File: engine/src/main/java/org/terasology/rendering/nui/asset/UIFormat.java [hunk 2]
```diff
@@ -196,8 +160,20 @@ public UIWidget deserialize(JsonElement json, Type typeOfT, JsonDeserializationC
             }
 
             // Deserialize normal fields.
+            Set<String> unknownFields = new HashSet<>();
+            for (Entry<String, JsonElement> entry : jsonObject.entrySet()) {
+                String name = entry.getKey();
+                if (!ID_FIELD.equals(name)
```
Comment (immortius) [inline engine/src/main/java/org/terasology/rendering/nui/asset/UIFormat.java]: Would consider putting these fields in a static immutable set or similar, and replacing with a call to contains(), but not a big deal with only 4 fields.

File: engine/src/main/resources/assets/ui/menu/selectModsScreen.ui (hunk 1 above)
Comment (msteiger) [inline engine/src/main/resources/assets/ui/menu/selectModsScreen.ui]: The previous one was wrong. `ColumnLayout` defines a custom serialized name for it.
